UnityProject/Assets/Scripts/Messages/Client/RequestShootMessage.cs

```diff
@@ -3,43 +3,49 @@
 using UnityEngine.Networking;
 
 /// <summary>
-/// Informs server of a shoot action
+/// Informs server of a shoot action (implied that the sender is the shooter)
 /// </summary>
 public class RequestShootMessage : ClientMessage {
 
 	public static short MessageType = (short)MessageTypes.RequestShootMessage;
-	public NetworkInstanceId Weapon;
-	public NetworkInstanceId ShotBy;
-	public Vector2 Direction;
-	public string BulletName;
+	/// <summary>
+	/// Position the player is shooting at (NOT the same as the final position - final position
+	/// may differ due to accuracy penalty)
+	/// </summary>
+	public Vector2 Target;
+	/// <summary>
+	/// Targeted damage zone
+	/// </summary>
 	public BodyPartType DamageZone;
+	/// <summary>
+	/// Whether this is a suicide shot
+	/// </summary>
 	public bool IsSuicideShot;
 
 
 	public override IEnumerator Process()
 	{
-		if (ShotBy.Equals(NetworkInstanceId.Invalid) || Weapon.Equals(NetworkInstanceId.Invalid)) {
+		if (SentBy.Equals(NetworkInstanceId.Invalid)) {
 			//Failfast
 			Logger.LogWarning($"Shoot request invalid, processing stopped: {ToString()}", Category.Firearms);
 			yield break;
 		}
 
 
-		yield return WaitFor(SentBy, ShotBy, Weapon);
-		Weapon wep = NetworkObjects[2].GetComponent<Weapon>();
-		wep.ServerShoot(NetworkObjects[1],Direction,BulletName, DamageZone, IsSuicideShot);
+		yield return WaitFor(SentBy);
+		//get the currently equipped weapon in the player's active hand
+		PlayerNetworkActions pna = NetworkObject.GetComponent<PlayerNetworkActions>();
+		Weapon wep = pna.GetActiveHandItem().GetComponent<Weapon>();
+		wep.ServerShoot(NetworkObject,Target, DamageZone, IsSuicideShot);
 	}
 
 	public static RequestShootMessage Send(GameObject weapon, Vector2 direction, string bulletName,
 	                                       BodyPartType damageZone, bool isSuicideShot, GameObject shotBy)
 	{
 		RequestShootMessage msg = new RequestShootMessage {
-			Weapon = weapon ? weapon.GetComponent<NetworkIdentity>().netId : NetworkInstanceId.Invalid,
-			Direction = direction,
-			BulletName = bulletName,
+			Target = direction,
```

Comment:
When I originally implemented shooting I created a ray cast on the client and captured the first point of collision and then sent the collision position to the server to determine direction there. This way it doesn't matter where the start point is (due to lag player could be further ahead on client then on the server) the bullet will always end up impacting the same spot. Lets see how this goes for a bit but in the future if there are too many inconsistencies then we should revert back to sending the end point of the trajectory and calculating direction via server.